internal/impl/mssqlserver/input_mssql_server_cdc.go

```diff
@@ -0,0 +1,385 @@
+	defer done()
+
+	if snapshot != nil {
+		if err := snapshot.prepare(ctx, r.tables); err != nil {
+			return fmt.Errorf("processing snapshot: %s", err)
+		}
+	}
+
+	ctStream := &changeTableStream{
+		logger:        r.logger,
+		trackedTables: make(map[string]changeTable, len(r.tables)),
+		cachedLSN:     cachedLSN,
+	}
+
+	if err := ctStream.verifyChangeTables(ctx, r.db, r.tables); err != nil {
+		return fmt.Errorf("verifying MS MSQL Server change tables: %s", err)
+	}
+
+	go func() {
+		ctx, _ = r.stopSig.SoftStopCtx(context.Background())
+		wg, ctx := errgroup.WithContext(ctx)
+		wg.Go(func() error { return r.readMessages(ctx, ctStream) })
+		wg.Go(func() error { return r.batchMessages(ctx) })
+
+		if err := wg.Wait(); err != nil && !errors.Is(err, context.Canceled) {
+			r.logger.Errorf("Error during Microsoft SQL Server CDC: %s", err)
+		} else {
+			r.logger.Info("Successfully shutdown Microsoft SQL Server CDC stream")
+		}
+		r.stopSig.TriggerHasStopped()
+	}()
+
+	return nil
+}
+
+func (r *msSqlServerCDCReader) batchMessages(ctx context.Context) error {
+	var nextTimedBatchChan <-chan time.Time
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-nextTimedBatchChan:
+			nextTimedBatchChan = nil
+			flushedBatch, err := r.batchPolicy.Flush(ctx)
+			if err != nil {
+				return fmt.Errorf("timed flush batch error: %w", err)
+			}
+			if err := r.flushBatch(ctx, r.checkpoint, flushedBatch); err != nil {
+				return fmt.Errorf("failed to flush periodic batch: %w", err)
+			}
+		case c := <-r.rawMessageEvents:
+			data, err := json.Marshal(c.Data)
+			if err != nil {
+				return fmt.Errorf("failure to marshal message: %w", err)
+			}
+			msg := service.NewMessage(data)
+			msg.MetaSet("table", c.Table)
+			if c.StartLSN != nil {
+				msg.MetaSet("start_lsn", string(c.StartLSN))
+			}
+			if c.EndLSN != nil {
+				msg.MetaSet("end_lsn", string(c.EndLSN))
+			}
+			msg.MetaSet("sequence_value", string(c.SequenceValue))
+			msg.MetaSet("operation", fmt.Sprint(c.Operation))
+			msg.MetaSet("update_mask", string(c.UpdateMask))
+			msg.MetaSet("command_id", strconv.Itoa(c.CommandID))
+
+			if r.batchPolicy.Add(msg) {
+				nextTimedBatchChan = nil
+				flushedBatch, err := r.batchPolicy.Flush(ctx)
+				if err != nil {
+					return fmt.Errorf("flush batch error: %w", err)
+				}
+				if err := r.flushBatch(ctx, r.checkpoint, flushedBatch); err != nil {
+					return fmt.Errorf("failed to flush batch: %w", err)
+				}
+			} else {
+				d, ok := r.batchPolicy.UntilNext()
+				if ok {
+					nextTimedBatchChan = time.After(d)
+				}
+			}
+		}
+	}
+}
+
+func (r *msSqlServerCDCReader) readMessages(ctx context.Context, stream *changeTableStream) error {
+	err := stream.readChangeTables(ctx, r.db, func(c *change) (LSN, error) {
+		// fmt.Printf("LSN=%x, CommandID=%d, SeqVal=%x, op=%d table=%s cols=%d\n", c.startLSN, c.commandID, c.seqVal, c.operation, c.table, len(c.columns))
+		r.rawMessageEvents <- MessageEvent{
```

Comment:
It's probably worth wrapping this in a select that detects shutdown and allows escape.